libs/neon-shmem/src/lib.rs

```diff
@@ -0,0 +1,376 @@
+            })
+        };
+
+        // The user data begins after the header
+        let data_ptr = unsafe { start_ptr.cast().add(HEADER_SIZE) };
+
+        Ok(ShmemHandle {
+            fd,
+            max_size: max_size.into(),
+            shared_ptr: shared,
+            data_ptr,
+        })
+    }
+
+    // return reference to the header
+    fn shared(&self) -> &SharedStruct {
+        unsafe { self.shared_ptr.as_ref() }
+    }
+
+    /// Resize the shared memory area. 'new_size' must not be larger than the 'max_size' specified
+    /// when creating the area.
+    ///
+    /// This may only be called from one process/thread concurrently. We detect that case
+    /// and return an Error.
+    pub fn set_size(&self, new_size: usize) -> Result<(), Error> {
+        let new_size = new_size + HEADER_SIZE;
+        let shared = self.shared();
+
+        if new_size > self.max_size {
+            panic!("size is greater than max_size");
+        }
+        assert_eq!(self.max_size, shared.max_size);
+
+        // "Lock" the area by setting the bit in 'current_size'
+        let mut old_size = shared.current_size.load(Ordering::Relaxed);
+        loop {
+            if (old_size & RESIZE_IN_PROGRESS) != 0 {
+                return Err(Error::new(
+                    "concurrent resize detected",
+                    Errno::UnknownErrno,
+                ));
+            }
+            match shared.current_size.compare_exchange(
+                old_size,
+                new_size,
+                Ordering::Relaxed,
+                Ordering::Relaxed,
+            ) {
+                Ok(_) => break,
+                Err(x) => old_size = x,
+            }
+        }
+
+        // Ok, we got the "lock".
+        //
+        // NB: If anything goes wrong, we must clear the bit!
+        let result = {
+            use std::cmp::Ordering::{Equal, Greater, Less};
+            match new_size.cmp(&old_size) {
+                Less => nix_ftruncate(&self.fd, new_size as i64).map_err(|e| {
+                    Error::new("could not shrink shmem segment, ftruncate failed: {e}", e)
+                }),
+                Equal => Ok(()),
+                Greater => nix_posix_fallocate(&self.fd, 0, new_size as i64).map_err(|e| {
+                    Error::new(
+                        "could not grow shmem segment, posix_fallcate failed: {e}",
+                        e,
+                    )
+                }),
+            }
+        };
+
+        // Unlock
+        shared.current_size.store(
+            if result.is_ok() { new_size } else { old_size },
+            Ordering::Relaxed,
+        );
+
+        result
+    }
+
+    /// Returns the current user-visible size of the shared memory segment.
+    pub fn current_size(&self) -> usize {
+        let total_current_size =
+            self.shared().current_size.load(Ordering::Relaxed) & !RESIZE_IN_PROGRESS;
+        total_current_size - HEADER_SIZE
+    }
+}
```

Comment:
We'll need to synchronize access to avoid out-of-bounds accesses with a stale `current_size`. I assume we'll do that via the hashmap? Worth calling that out here.

Reply:
Right, it's the caller's responsibility to ensure that you don't access a region that's concurrently truncated away. I'll add a note here.